pkg/netpol/connlist/internal/ingressanalyzer/ingress_analyzer.go

```diff
@@ -0,0 +1,408 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/common"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/eval"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/logger"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/scan"
+)
+
+const (
+	//  The actual ingress controller pod is usually unknown and not available in the input resources for the analysis.
+	// IngressPodName and IngressPodNamespace are used to represent  that pod with those placeholder values for name and namespace.
+	IngressPodName      = "ingress-controller"
+	IngressPodNamespace = "ingress-controller-ns"
+)
+
+type serviceInfo struct {
+	// used to populate routesToServicesMap and k8sIngressToServicesMap with their target services info
+	serviceName string
+	servicePort intstr.IntOrString // service port name or port number
+}
+
+type peersAndPorts struct {
+	// used to populate the servicesToPortsAndPeersMap Map with its ports and selected peers
+	peers []eval.Peer
+	ports []corev1.ServicePort
+}
+
+// IngressAnalyzer provides API to analyze Ingress/Route resources, to allow inferring potential connectivity
+// from ingress-controller to pods in the cluster
+type IngressAnalyzer struct {
+	logger                     logger.Logger
+	pe                         *eval.PolicyEngine                  // holds the workload peers relevant to the analysis
+	servicesToPortsAndPeersMap map[string]map[string]peersAndPorts // map from namespace to map from service name to its ports and
+	// its selected workloads
+	routesToServicesMap     map[string]map[string][]serviceInfo // map from namespace to map from route name to its target service
+	k8sIngressToServicesMap map[string]map[string][]serviceInfo // map from namespace to map from k8s ingress object name to
+	// its target services
+}
+
+// NewIngressAnalyzerWithObjects returns a new IngressAnalyzer with relevant objects
+func NewIngressAnalyzerWithObjects(objects []scan.K8sObject, pe *eval.PolicyEngine, l logger.Logger) (*IngressAnalyzer, error) {
+	ia := &IngressAnalyzer{
+		servicesToPortsAndPeersMap: make(map[string]map[string]peersAndPorts),
+		routesToServicesMap:        make(map[string]map[string][]serviceInfo),
+		k8sIngressToServicesMap:    make(map[string]map[string][]serviceInfo),
+		logger:                     l,
+		pe:                         pe,
+	}
+
+	var err error
+	for _, obj := range objects {
+		switch obj.Kind {
+		case scan.Service:
+			err = ia.mapServiceToPeers(obj.Service)
+		case scan.Route:
+			ia.mapRouteToServices(obj.Route)
+		case scan.Ingress:
+			ia.mapk8sIngressToServices(obj.Ingress)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	return ia, err
+}
+
+// IsEmpty returns whether there are no services to consider for Ingress analysis
+func (ia *IngressAnalyzer) IsEmpty() bool {
+	return len(ia.routesToServicesMap) == 0 && len(ia.k8sIngressToServicesMap) == 0
+}
```

Comment:
why is the analysis required if there are Route/Ingress objects, but no Service objects?

Reply:
if there are no service objects, the ia.routesToServicesMap and ia.k8sIngressToServicesMap will be empty. 
it there are no services matching the Route/Ingress object, then it will not have an entry in the relevant map

Reply:
will update the return statement